### ISSUE TITLE
Support custom errors

### ISSUE THREAD

**Issue by seanyoung** [hyperledger/solang#698]
Ethereum Solidity 0.8 introduces a new syntax for custom errors.

sema changes:
- Using the `error Foo(int,string)` type syntax, custom errors can be declared. These need resolving in sema, which should be identical to event definition parsing
- Resolve `revert Foo(12, "bar")` type syntax. Again this should be very similar to `emit` resolving.
- The error defined in another contract be used using `revert ContractX.Foo(12, "bar")`
- The error can be imported from another contract via an import type:
```
// foo.sol
contract X  { error Foo(int,string); }
}
// bar.sol
import "foo.sol" as I;

contract A {
    function x() public { revert I.X.Foo(1, ""); }
}
```
- import types can imported again, so it's possible to have:
```
contract A {
    function x() public { revert I.a.b.c.X.Foo(1, ""); }
}
```

codegen changes:
- The should be a new Revert instruction for the cfg which looks like emit
- All the passes should be able to use expressions in the revert instruction, just like emit

emit changes:
 - Generate llvm for new revert instruction. This should be similar to emit, but now terminates the contract

try-catch support for custom errors:
 - The try-catch syntax can catch any number of custom errors. This needs support too

**Comment by xermicus** [comment]
Found elsewhere but seems related here: 

```solidity
pragma solidity ^0.8.0;

contract Implementation {
    function foo(bool ok, string s) public pure {
        require(ok, s);
    }
}

contract A {
    address init;

    constructor(address init_addr) {
        init = init_addr;
    }

    function f(bytes calldata _calldata) public {
        (bool success, bytes memory error) = init.delegatecall(_calldata);
        if (!success) {
            if (error.length > 0) {
                revert(string(error));
            } else {
                revert("aa");
            }
        }
    }
}
```

This isn't displayed correctly:

```
runtime_error: 56a5f41bf5df82020a2efe5ee1894169a0f555d980038a809ed1eb469b8d64a4000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.... 
```

Also this makes a nice testcase.

**Comment by xermicus** [comment]
Closed by #1516
